src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala

```diff
@@ -6,68 +6,128 @@ import is.hail.utils._
   val partitionKey: Array[String], val kType: TStruct,
-  // rangeBounds is partition max, sorted ascending
-  // rangeBounds: Array[pkType]
+  // rangeBounds: Array[Interval[pkType]]
+  // rangeBounds is interval containing all partition keys within a partition
   val rangeBounds: UnsafeIndexedSeq) extends Partitioner {
-  require((numPartitions == 0 && rangeBounds.isEmpty) || numPartitions == rangeBounds.length + 1,
-    s"nPartitions = $numPartitions, ranges = ${ rangeBounds.length }")
+  val numPartitions: Int = rangeBounds.length
 
   val (pkType, _) = kType.select(partitionKey)
+  val pkIntType = TInterval(pkType)
+  val rangeBoundsType = TArray(pkIntType)
 
-  val pkKFieldIdx: Array[Int] = partitionKey.map(n => kType.fieldIdx(n))
-  val pkKOrd: UnsafeOrdering = OrderedRVDType.selectUnsafeOrdering(pkType, (0 until pkType.size).toArray, kType, pkKFieldIdx)
-
-  val rangeBoundsType = TArray(pkType)
   assert(rangeBoundsType.typeCheck(rangeBounds))
 
+  require(rangeBounds.isEmpty || rangeBounds.zip(rangeBounds.tail).forall { case (left: Interval, right: Interval) =>
+    !left.overlaps(pkType.ordering, right) && pkType.ordering.compare(left.start, right.start) <= 0
+  })
+
+  val rangeTree: IntervalTree[Int] = new IntervalTree[Int](IntervalTree.fromSorted(pkType.ordering,
+    Array.tabulate[(BaseInterval, Int)](numPartitions) { i =>
+      (rangeBounds(i).asInstanceOf[Interval], i)
+    }, 0, rangeBounds.size))
+
+  val pkKFieldIdx: Array[Int] = partitionKey.map(n => kType.fieldIdx(n))
+  val pkKOrd: ExtendedOrdering = OrderedRVDType.selectExtendedOrdering(pkType, (0 until pkType.size).toArray, kType, pkKFieldIdx)
+
   val ordering: Ordering[Annotation] = pkType.ordering.toOrdering
-  require(rangeBounds.isEmpty || rangeBounds.zip(rangeBounds.tail).forall { case (left, right) => ordering.compare(left, right) < 0 })
 
   def region: Region = rangeBounds.region
 
-  def loadElement(i: Int): Long = rangeBoundsType.loadElement(rangeBounds.region, rangeBounds.aoff, rangeBounds.length, i)
+  def loadElement(i: Int): Long = rangeBoundsType.loadElement(region, rangeBounds.aoff, rangeBounds.length, i)
+
+  def loadStart(i: Int): Long = pkIntType.loadStart(region, loadElement(i))
+
+  def loadEnd(i: Int): Long = pkIntType.loadStart(region, loadElement(i))
 
-  // return the smallest partition for which key <= max
   // pk: Annotation[pkType]
   def getPartitionPK(pk: Any): Int = {
     assert(pkType.typeCheck(pk))
-
-    val part = BinarySearch.binarySearch(numPartitions,
-      // key.compare(elem)
-      i =>
-        if (i == numPartitions - 1)
-          -1 // key.compare(inf)
-        else
-          ordering.compare(pk, rangeBounds(i)))
-    part
+    val part = rangeTree.queryValues(pkType.ordering, pk)
+    part.length match {
+      case 0 =>
+        if (pkType.ordering.gt(pk, rangeTree.root.get.maximum))
+          numPartitions - 1
+        else {
+          assert(pkType.ordering.lt(pk, rangeTree.root.get.minimum))
```

Comment:
Right now you require the form of the intervals to be [a, b], (b, c], (c, d], ..., right?  Otherwise, I think this is wrong.  If so, require that in the partitioner constructor so we know to fix this.